src/components/containers/TraceAccordion.js

```diff
@@ -10,78 +10,155 @@ import {Tab, Tabs, TabList, TabPanel} from 'react-tabs';
 const TraceFold = connectTraceToPlot(PlotlyFold);
 
 class TraceAccordion extends Component {
-  render() {
-    const {data = [], localize: _} = this.context;
-    const {
-      canAdd,
-      canGroup,
-      children,
-      messageIfEmptyFold,
-      excludeFits,
-    } = this.props;
+  renderGroupedTraceFolds(groupedTraces) {
+    if (this.props.useFullData) {
+      const dataArrayPositionsByTraceType = {};
+      const fullDataArrayPositionsByTraceType = {};
 
-    // we don't want to include analysis transforms when we're in the create panel
-    const filteredData = data.filter(t => {
-      if (excludeFits) {
-        return !(t.transforms && t.transforms.every(tr => tr.type === 'fit'));
-      }
-      return true;
-    });
+      Object.keys(groupedTraces).forEach(traceType => {
+        const dataIndexes = [];
+        const fullDataIndexes = [];
+
+        this.context.fullData.forEach((trace, fullDataIndex) => {
+          if (trace.type === traceType && trace._expandedIndex) {
+            fullDataIndexes.push(trace._expandedIndex);
+          }
+          if (trace.type === traceType && !trace._expandedIndex) {
+            fullDataIndexes.push(fullDataIndex);
+          }
+          if (trace.type === traceType) {
+            dataIndexes.push(trace.index);
+          }
+        });
+
+        dataArrayPositionsByTraceType[traceType] = dataIndexes;
+        fullDataArrayPositionsByTraceType[traceType] = fullDataIndexes;
+      });
 
-    const individualTraces =
-      filteredData.length &&
-      filteredData.map((d, i) => {
+      return Object.keys(groupedTraces).map((traceType, index) => {
         return (
           <TraceFold
-            key={i}
-            traceIndexes={[i]}
-            canDelete={canAdd}
-            messageIfEmpty={messageIfEmptyFold}
+            key={index}
+            traceIndexes={dataArrayPositionsByTraceType[traceType]}
+            name={traceType}
+            fullDataArrayPosition={fullDataArrayPositionsByTraceType[traceType]}
           >
-            {children}
+            {this.props.children}
           </TraceFold>
         );
       });
+    }
 
-    if (canAdd) {
-      const addAction = {
-        label: _('Trace'),
-        handler: ({onUpdate}) => {
-          if (onUpdate) {
-            onUpdate({
-              type: EDITOR_ACTIONS.ADD_TRACE,
-            });
-          }
-        },
-      };
+    return Object.keys(groupedTraces).map((traceType, index) => {
       return (
-        <PlotlyPanel addAction={addAction}>
-          {individualTraces ? individualTraces : null}
-        </PlotlyPanel>
+        <TraceFold
+          key={index}
+          traceIndexes={groupedTraces[traceType]}
+          name={traceType}
+        >
+          {this.props.children}
+        </TraceFold>
       );
+    });
+  }
+
+  renderIndividualTraceFolds(filteredTraces, filteredTracesFullDataPositions) {
+    if (this.props.useFullData) {
```

Comment:
Good refactoring! It shows that there's basically no reuse in this function any more, so we should go the next step and split it in half: `renderTraceFolds` for the non-`canGroup` case and `renderUnGroupedTraceFolds` for the `canGroup`-but-ungrouped case.